tools/oeedger8r/Emitter.ml

```diff
@@ -126,55 +127,70 @@ let write_file (content : string list) (filename : string) (dir : string) =
-  let type_expr =
-    match ptype with
-    | PTPtr (_, ptr_attr) ->
-        if ptr_attr.pa_isptr then sprintf "*(%s)0" base_t else base_t
-    | _ -> base_t
-  in
-  (* Convert an attribute to string. *)
-  let attr_value_to_string av =
-    match av with
-    | None -> ""
-    | Some (ANumber n) -> string_of_int n
-    | Some (AString s) -> argstruct ^ s
-    (* another parameter name *)
+let get_type_expr ptype =
+  (* Get the base type of the parameter. That is, yield its [atype],
+     unless it is a pointer, in which case decompose and yield the
+     [atype] the pointer points to. *)
+  let param_atype =
+    let a = get_param_atype ptype in
+    match a with Ptr p -> p | _ -> a
   in
-  let pa_size_to_string pa =
-    let c = attr_value_to_string pa.ps_count in
-    if c <> "" then sprintf "(%s * sizeof(%s))" c type_expr
-    else attr_value_to_string pa.ps_size
-  in
-  let decl_size_to_string (ptype : parameter_type) (d : declarator) =
-    let dims = List.map (fun i -> "[" ^ string_of_int i ^ "]") d.array_dims in
-    let dims_expr = String.concat "" dims in
-    sprintf "sizeof(%s%s)" type_expr dims_expr
+  let tystr = get_tystr param_atype in
+  match ptype with
+  | PTPtr (_, ptr_attr) when ptr_attr.pa_isptr -> sprintf "*(%s)0" tystr
+  | _ -> tystr
+
+let attr_value_to_string argstruct = function
+  | None -> None
+  | Some (ANumber n) -> Some (string_of_int n)
+  | Some (AString s) -> Some (argstruct ^ s)
+
+(** For a parameter, get its size expression. *)
+let oe_get_param_size (ptype, decl, argstruct) =
+  let type_expr = get_type_expr ptype in
+  let get_ptr_or_decl_size (p : ptr_size) (d : declarator) =
+    match attr_value_to_string argstruct p.ps_count with
+    | Some s -> sprintf "(%s * sizeof(%s))" s type_expr
+    | None -> (
+      match attr_value_to_string argstruct p.ps_size with
+      | Some s -> s
+      | None ->
+          let dims =
+            List.map (fun i -> "[" ^ string_of_int i ^ "]") d.array_dims
+          in
+          sprintf "sizeof(%s%s)" type_expr (String.concat "" dims) )
   in
   match ptype with
   | PTPtr (_, ptr_attr) ->
-      let pa_size = pa_size_to_string ptr_attr.pa_size in
-      (* Compute declared size *)
-      let decl_size = decl_size_to_string ptype decl in
       if ptr_attr.pa_isstr then
         argstruct ^ decl.identifier ^ "_len * sizeof(char)"
       else if ptr_attr.pa_iswstr then
         argstruct ^ decl.identifier ^ "_len * sizeof(wchar_t)"
-      else if (* Prefer size attribute over decl size *)
-              pa_size = "" then decl_size
-      else pa_size
+      else get_ptr_or_decl_size ptr_attr.pa_size decl
+  (* Values have no marshalling size. *)
   | _ -> ""
 
+(** For a parameter, get its count expression. *)
+let oe_get_param_count (ptype, decl, argstruct) =
+  let type_expr = get_type_expr ptype in
+  let get_ptr_or_decl_count (p : ptr_size) (d : declarator) =
+    match attr_value_to_string argstruct p.ps_count with
+    | Some s -> s
```

Comment:
Action item: Can you follow up if handling count & size scenario would require changing this function

Reply:
Ditto